freeipa/src/test/java/com/sequenceiq/freeipa/service/stack/FreeIpaUpgradeCcmServiceTest.java

```diff
@@ -72,13 +94,80 @@ void upgradeCcmTestWhenAvailableAndOperationRunning() {
         OperationStatus result = underTest.upgradeCcm(ENVIRONMENT_CRN, ACCOUNT_ID);
 
         assertThat(result).isSameAs(operationStatus);
-        // TODO verify that flow has been triggered successfully; see CB-14571
+        ArgumentCaptor<UpgradeCcmFlowChainTriggerEvent> eventCaptor = ArgumentCaptor.forClass(UpgradeCcmFlowChainTriggerEvent.class);
+        verify(flowManager).notify(eq(UPGRADE_CCM_CHAIN_TRIGGER_EVENT), eventCaptor.capture());
+        UpgradeCcmFlowChainTriggerEvent event = eventCaptor.getValue();
+        assertThat(event.getOldTunnel()).isEqualTo(Tunnel.CCM);
+        assertThat(event.getOperationId()).isEqualTo(OPERATION_ID);
+        assertThat(event.getResourceId()).isEqualTo(STACK_ID);
+    }
+
+    @EnumSource(value = Tunnel.class, names = { "DIRECT", "CLUSTER_PROXY"}, mode = EnumSource.Mode.INCLUDE)
+    @ParameterizedTest
+    void upgradeCcmTestWhenAvailableButIncorrectTunnelType(Tunnel tunnel) {
+        Stack stack = createStack(Status.AVAILABLE);
+        stack.setTunnel(tunnel);
+        when(stackService.getByEnvironmentCrnAndAccountIdWithListsAndMdcContext(ENVIRONMENT_CRN, ACCOUNT_ID)).thenReturn(stack);
+
+        assertThatThrownBy(() -> underTest.upgradeCcm(ENVIRONMENT_CRN, ACCOUNT_ID)).isInstanceOf(BadRequestException.class);
     }
 
-    // TODO add test for AvailableAndOperationRunningAndFlowStartFailure; see CB-14571
+    @Test
+    void upgradeCcmTestWhenAlreadyUpgraded() {
+        Stack stack = createStack(Status.AVAILABLE);
+        stack.setTunnel(Tunnel.CCMV2_JUMPGATE);
+        when(stackService.getByEnvironmentCrnAndAccountIdWithListsAndMdcContext(ENVIRONMENT_CRN, ACCOUNT_ID)).thenReturn(stack);
+
+        OperationStatus operationStatus = underTest.upgradeCcm(ENVIRONMENT_CRN, ACCOUNT_ID);
+        assertThat(operationStatus.getStatus()).isEqualTo(OperationState.COMPLETED);
+    }
+
+    @Test
+    void upgradeCcmTestWhenNotEntitledFromCCMv1() {
+        Stack stack = createStack(Status.AVAILABLE);
+        stack.setTunnel(Tunnel.CCM);
+        lenient().when(entitlementService.ccmV1ToV2JumpgateUpgradeEnabled(any())).thenReturn(false);
+
+        when(stackService.getByEnvironmentCrnAndAccountIdWithListsAndMdcContext(ENVIRONMENT_CRN, ACCOUNT_ID)).thenReturn(stack);
+
+        assertThatThrownBy(() -> underTest.upgradeCcm(ENVIRONMENT_CRN, ACCOUNT_ID)).isInstanceOf(BadRequestException.class);
+    }
+
+    @Test
+    void upgradeCcmTestWhenNotEntitledFromCCMv2() {
+        Stack stack = createStack(Status.AVAILABLE);
+        stack.setTunnel(Tunnel.CCMV2);
+        lenient().when(entitlementService.ccmV2ToV2JumpgateUpgradeEnabled(any())).thenReturn(false);
```

Comment:
I wonder if this will indeed effective. `setUp()` already has a stubbing for this invocation with a result of `true`.

Reply:
it is effective